### ISSUE TITLE
Can not search in older releases

### ISSUE THREAD

**Issue by flip111** [commercialhaskell/stackage-server#326]
Reproduce

* Go to https://www.stackage.org/
* Click on [LTS 14.27 for ghc-8.6.5](https://www.stackage.org/lts-14.27)
* In search box type `exceptions` and click Search

Note the following:
* URL is set to: `https://www.stackage.org/lts-14.27/hoogle?q=exceptions`
* On the page there is the text `Within [LTS Haskell 22.11 (ghc-9.6.4)](https://www.stackage.org/lts-22.11)`

Continue
* Click on [package exceptions](https://www.stackage.org/lts-22.11/package/exceptions)

Note the following:
Getting package information for exceptions in LTS-22.11

**Comment by chreekat** [comment]
fwiw we have a bunch of disk space we could use for caching hoogle databases now. But we'd still need some sort of LRU cache rather than keeping all of them local.

**Comment by chreekat** [comment]
One could even imagine other solutions where the database is never even brought onto the local disk. I imagine it might be way too slow to always fetch from the bucket, of course. But these servers also have a lot of memory. The LRU could even exist in ram?

Notwithstanding these musings, I suspect it'll still be better to use the local disk as the cache.

**Comment by flip111** [comment]
I looked through the code and build the hoogle database locally. What i saw was this:

```
stack hoogle --rebuild
Minimum version is hoogle-5.0. Found acceptable hoogle-5.0.18.4 in your index, requiring its installation.
```

this uses lts-23.17 (what was configured in my global project)

I'm assuming (but this might be incorrect) that the hoogle database being built on everything available in the stackage release.

```
la -B .stack/global-project/.stack-work/hoogle/x86_64-linux-tinfo6/592d9a894609f023ea179dffa09efa4564e9ac2b2c5f10da24af78b72616d817/9.8.4/database.hoo
```
This gives 9,776,531 bytes (9.8 MB)

The database is not read as a file, de-serialized and then converted into data structures. Instead it's directly mapped into the memory and the search is build to read directly from the memory stream with de-serializing or further moving data around https://github.com/ndmitchell/hoogle/blob/e83646ee1b62eb724def6c838a34017b60fc519a/src/General/Store.hs#L170

When i look at all of the versions on the front-page of stackage https://www.stackage.org/ there are 33 listed versions (excluding latest nightly). Of course there are more versions into existence, but those 34 are a good beginning and probably the most important ones. So to have that all mapped into memory is 332,402,054 bytes (332.4MB). Maybe the server capacity is large enough to keep them loaded in memory permanently. But anyway when using a least recently used cache in practice it would probably be those same 34 versions anyway, so i don't know if it makes sense to keep any version of the hoogle database pinned permanently.

For the least recently used cache i saw these two packages:
* https://github.com/chowells79/lrucache
* https://github.com/hverr/haskell-cache

they are pretty old though, and probably the LRU mechanism itself doesn't have to be very optimized, since it's more about loading and unloading the files to memory anyway.

there is also the option of Amazon ElastiCache with Redis, i imagine that way there is more RAM available, or it can be shared between different pods. However i see some tradeoffs as well
* increased complexity in infrastructure setup
* Application becomes tied to specific AWS stuff. For example through this library https://hackage.haskell.org/package/amazonka-elasticache
* Potential of cost-lowering. If there are more instances of the application then a shared ElastiCache can save costs. Potential cost-increase due to paying for ElastiCache. So needs some study (extra effort) to find out if that is worth it.
* Though potentially easier scaling through the AWS dashboards rather than changing the configuration file of an application

Then there could also be a second mechanism where the file is copied from the S3 to the local file storage since S3 can be slow. I guess if the space is there anyway and it's unused then this can be used. But of course for the most pleasant experience for users you want to have things loaded into a cache anyway.

Directly loading from S3 to memory skipping local files is possible but also requires some modification of the stackage-server code as it requires a local path currently.

I think it should work in such a way that if you just start the stackage-server app normally that it just memory maps that one single database as it does now. But it should take some optional configuration file that lets you choose the have an LRU cache of either up to X versions of a database or up to Y megabytes of RAM. This might require some modifications in the hoogle library as well, or we have to copy over code into stackage-server that now lives in hoogle to modify it. Also configuration to have local files LRU with a max versions / max file-space would be cool too.

It's hard to develop these functionalities with an AWS sandbox though since some functionality is specifically designed to work with AWS such as S3 buckets.

**Comment by flip111** [comment]
@chreekat what are the specs of the AWS EC2 instances and current pricing and quantity?

**Comment by chreekat** [comment]
@flip111 it's not on AWS. :) The new box is a beefy piece of hardware living in a colo. There's just one of them reserved for Stackage, but it's way more than we currently need.

And the "S3 compatible" object storage I mentioned earlier is actually Cloudflare's R2 service.

**Comment by flip111** [comment]
@chreekat what's the best plan of attack here? a pull request to stackage-server with caching mechanisms and whatnot?

**Comment by chreekat** [comment]
Yeah, that might be fine. Let's figure out exactly how we want to implement it, though.

For discussion, here's what my draft attempt looks like:

https://github.com/haskellfoundation/hf-infrastructure/pull/15/files

In there, you'll find a Bash one-liner run as a systemd timer. It simply removes all but the 5 most recently-created snapshots. It's a draft, because I noticed I need to handle LTS and Nightly snapshots individually. (Stackage still uses both at the same time.)

If we wanted to turn that into removing the oldest-*used* databases, we could use `ls -1tr --time=access`.

And if we wanted to keep around more of the databases, we could bump the number from 5 to, say, 500 total (250 LTS and 250 Nightly).

Finally, if we wanted to make them *available*, we would revert 68fa14a4 .

Relying on access time could potentially be fragile. The access time of the databases could get modified outside of Stackage by random unexpected actions, including things done manually on the shell. But the consequence of that would be, at worst, the deletion of snapshots that people still want to use -- in which case, Stackage would just re-download them. So I think that would probably be fine.

So like I said, that just for discussion. Do you have any other idea you'd like to try?

**Comment by chreekat** [comment]
1. stackage-server checks for the presence of the file on disk every time the endpoint is hit. There is, however, a race condition between checking that it exists, and reading the file.
2. There are pros and cons to keeping the queues separate - will discuss more.
3. I'd rather use a fixed usage limit, instead of a percentage of available space, because it's more predictable.
4. I'll discuss pros and cons of having stackage-server manage it internally below
5. I don't know the response times, and honestly I'm not extremely worried about it.
6. Yeah, that's why I said we would need to revert 68fa14a. That was the commit that started shunting searches to always use the latest Nightly or latest LTS, regardless of which snapshot page the search originate from. There was also some commit that added the text "Note that Stackage only displays results for the latest LTS and Nightly snapshot. [Learn more.](https://github.com/commercialhaskell/stackage-server/issues/305)" that we'd need to revert.

![Image](https://github.com/user-attachments/assets/089f5a7e-3a1a-4e87-b3ea-09035de84876)

### Keeping queues separate?

My concern here is that access patterns are dominated by bots, and there are a lot more Nightly snapshots (2809) than LTS snapshots (562). If bots keep crawling all the Nightlies such that LTS snapshots get evicted, that would be a bad thing.

It might even make sense to just not evict LTSs at all.

### Mantain the queue in stackage-server? 

Pros:

1. Simpler deployment needs
2. No race condition accessing the file _if implemented correctly_. But I believe the race condition could be fixed by just retrying the download on read failure, anyway.

Cons:

1. A lot simpler to implement. systemd is already a full solution to scheduling and logging, and the script is already well-behaved under service restarts or system reboots.

**Overall**, I'd be happy doing it either way.

**Comment by flip111** [comment]
@chreekat do you need any help with implementation or testing?

**Comment by chreekat** [comment]
Yes please! It would be great if you could contribute.

**Comment by flip111** [comment]
@chreekat here is an updated script. Do you have any other ideas on how i could contribute?

```bash
#!/usr/bin/env bash
#
# stackage-hoogle-cleanup.sh - Cleanup script for Stackage Hoogle database files
#
# This script maintains the Hoogle databases for Stackage by:
# 1. Preserving specific LTS versions we want to keep
# 2. Keeping the most recently accessed LTS versions up to a configured total
# 3. Keeping the most recently accessed Nightly versions up to a configured total
# 4. Cleaning up empty directories after removal
#
# Usage: stackage-hoogle-cleanup.sh
#

# Enable strict mode
set -euo pipefail

# Enable debug mode if DEBUG is set to true
[[ "${DEBUG:-}" == "true" ]] && set -x

# Configuration
readonly LTS_VERSIONS_TO_KEEP=250
readonly NIGHTLY_VERSIONS_TO_KEEP=250
readonly HOOGLE_DIR="/home/stackage-server/hoogle"

# Define the list of LTS versions we want to keep explicitly
readonly KEEP_LTS=(
  "lts-23.19" # ghc-9.8.4
  "lts-22.43" # ghc-9.6.6
  "lts-22.27" # ghc-9.6.5
  "lts-22.20" # ghc-9.6.4
  "lts-22.6"  # ghc-9.6.3
  "lts-21.25" # ghc-9.4.8
  "lts-21.21" # ghc-9.4.7
  "lts-21.11" # ghc-9.4.6
  "lts-21.7"  # ghc-9.4.5
  "lts-20.26" # ghc-9.2.8
  "lts-20.24" # ghc-9.2.7
  "lts-20.12" # ghc-9.2.6
  "lts-20.11" # ghc-9.2.5
  "lts-19.33" # ghc-9.0.2
  "lts-18.28" # ghc-8.10.7
  "lts-18.8"  # ghc-8.10.6
  "lts-18.6"  # ghc-8.10.4
  "lts-17.2"  # ghc-8.10.3
  "lts-16.31" # ghc-8.8.4
  "lts-16.11" # ghc-8.8.3
  "lts-15.3"  # ghc-8.8.2
  "lts-14.27" # ghc-8.6.5
  "lts-13.19" # ghc-8.6.4
  "lts-13.11" # ghc-8.6.3
  "lts-12.26" # ghc-8.4.4
  "lts-12.14" # ghc-8.4.3
  "lts-11.22" # ghc-8.2.2
  "lts-9.21"  # ghc-8.0.2
  "lts-7.24"  # ghc-8.0.1
  "lts-6.35"  # ghc-7.10.3
  "lts-3.22"  # ghc-7.10.2
  "lts-2.22"  # ghc-7.8.4
  "lts-0.7"   # ghc-7.8.3
)

# Calculate how many additional LTS versions to keep beyond the hardcoded ones
readonly ADDITIONAL_LTS_TO_KEEP=$((LTS_VERSIONS_TO_KEEP - ${#KEEP_LTS[@]}))

# Function to clean up temporary files
cleanup() {
  if [[ -d "${TEMP_DIR:-}" ]]; then
    rm -rf "${TEMP_DIR}"
  fi
}

main() {
  # Ensure we're in the right directory
  cd "${HOOGLE_DIR}" || {
    printf "Error: Cannot change to directory %s\n" "${HOOGLE_DIR}" >&2
    return 1
  }

  # Create temporary directory for our files
  local TEMP_DIR
  TEMP_DIR=$(mktemp -d) || {
    printf "Error: Failed to create temporary directory\n" >&2
    return 1
  }
  
  # Set up cleanup trap
  trap cleanup EXIT

  # Process LTS versions
  process_lts_versions
  
  # Process Nightly versions
  process_nightly_versions
  
  # Clean up empty directories
  find . -type d -empty -delete
}

process_lts_versions() {
  local keep_files="${TEMP_DIR}/keep_files.txt"
  local all_files="${TEMP_DIR}/all_lts_files.txt"
  local delete_candidates="${TEMP_DIR}/lts_delete_candidates.txt"
  
  # First, create a list of files that match our keep list
  for version in "${KEEP_LTS[@]}"; do
    find "./lts-"* -type f -path "./${version}/*" -print0 >> "${keep_files}" 2>/dev/null || true
  done

  # Create a list of all LTS files
  find ./lts-* -type f -print0 > "${all_files}" 2>/dev/null || true

  # Create a list of files to consider for deletion (excluding our keep list)
  comm -0 -23 <(sort -z "${all_files}") <(sort -z "${keep_files}") > "${delete_candidates}"

  # From the deletion candidates, keep the most recently accessed files
  # and delete the rest
  if [[ -s "${delete_candidates}" ]]; then
    xargs -0 -r stat --format="%X %n" < "${delete_candidates}" 2>/dev/null | 
      sort -rn | 
      tail -n +$((ADDITIONAL_LTS_TO_KEEP + 1)) | 
      cut -d' ' -f2- | 
      xargs -r rm -f
  fi
}

process_nightly_versions() {
  # Process Nightly versions (keep NIGHTLY_VERSIONS_TO_KEEP most recently accessed)
  find ./nightly-* -type f -printf "%A@\t%p\0" 2>/dev/null |
    sort -z -n | 
    head -z -n -"${NIGHTLY_VERSIONS_TO_KEEP}" | 
    cut -z -f2- | 
    xargs -0 -r rm -f
}

# Run the main function
main "$@"
```

**Comment by chreekat** [comment]
The version I came up with is 

```bash
## Keep 250 most recently accessed databases for both LTS and Nightly separately.

cd "/home/${srvName}/hoogle"
for pattern in "./lts-*" "./nightly-*"; do
  find . -type f -path "$pattern" -printf "%A@\t%p\0" |
    sort -z -n | head -z -n -250 | cut -z -f2- | xargs -0 -r rm -f
done

# Clean up empty directories
find . -type d -empty -delete
```

The problem with bash is that you never know if it works until you try it. Less code is just always better. It looks like you've hardcoded the list of LTSs to keep, but then also you've checked to see if they exist, and silently ignore missing ones. It just adds a lot that needs to be evaluated and understood. When you zoom out and look at the whole system, there's already plenty of complexity. A small system like this is a lot more helpful if it's kinda stupid but extremely easy to understand.

Anyway, would you like to open a pull request that adds something like this? Well, I'll be honest -- I already have something basically ready. Maybe you'd like to review it once I've submitted it?

You could also open a PR with the revert commit we discussed. I haven't touched that yet.

**Comment by flip111** [comment]
> It looks like you've hardcoded the list of LTSs to keep

Yes because of `I'll just point out that trying to figure out which versions are shown on the front page is a lot of extra work, and may not be necessary.`

> and silently ignore missing ones

Missing ones should be re-downloaded, this script is for deleting

> It just adds a lot that needs to be evaluated and understood.

Only the minimum was added to meet the requirements, using good bash practices. Of course we can do codegolf on the bash script and make it look visually smaller.

> When you zoom out and look at the whole system, there's already plenty of complexity.

That would be a reason to built this into stackage-server directly at least then we get to enjoy haskell's type safety and whatnot.

> A small system like this is a lot more helpful if it's kinda stupid but extremely easy to understand.

You can do it with your bash script if you prefer that and revert the git commit and we are done?

**Comment by chreekat** [comment]
Let's get back to this next week - I need to finish up the migration of Stackage from Equinix to OpenCape right now!

**Comment by flip111** [comment]
Ok take your time @chreekat  :)

**Comment by chreekat** [comment]
Tell you what -- I still like my version better, but if you want to do it your way and open a PR, I would accept it!

**Comment by flip111** [comment]
I only made that version because i thought yours wasn't finished, you can take yours it's fine!

**Comment by chreekat** [comment]
The thing is, you might have more time to finish yours than I have to finish mine. 😂 My Stackage priorities are still about business continuity and disaster recovery -- setting up backups, following up on the migration, etc

**Comment by flip111** [comment]
What do you mean with finish though? What's not finished about your script?

**Comment by juhp** [comment]
(I personally feel just supporting major LTS versions would be sufficient or a good improvement)

**Comment by chreekat** [comment]
@flip111 yes, well, as you may have noticed, my availability may be insufficient to get it over the line. 🙈 All my bandwidth is currently spent on operational stuff atm. So I wanted to let you take it on if you're up for it.

**Comment by flip111** [comment]
@chreekat  that's understandable no worries. I'm not sure now though what it would require to get it over the line. I drafted a plan and made a script before and then you had some comments on it that seemed it was not completely satisfactory yet. So what is it we need concretely? Your script but then in a PR, my script in a PR? the commit reverse in a PR? or something else? I feel like we missed something or so, because just sending on those PR's with code that is already there seems trivial to me. Perhaps some testing or working with the infrastructure?

**Comment by chreekat** [comment]
What I was trying to say is that your script would be fine if you want to open a PR with it.

**Comment by drlkf** [comment]
Hello! I stumbled upon this when searching how to use Hoogle on older snapshots, I suppose it's still in progress? Can I help somehow?

**Comment by flip111** [comment]
@drlkf  yes look here https://github.com/haskellfoundation/hf-infrastructure/pull/15 implement something like this. You can let some AI read the discussion in this issue and come up with some proposals on how to move this forward
